### ISSUE TITLE
Sync istio config changes made by kiali to a repo

### ISSUE THREAD

**Issue by lucasmellos** [kiali/kiali#5089]
It would be nice to have a way to sync changes made by Kiali GUI to a remote repo. 
Let's say that you create a new request routing rule or edit some of these rules e.g. (VirtualService, Gateway, etc) and you don't want to lose the sync between your source code in your repo to the changes applied using Kiali GUI.

**Comment by jshaughn** [comment]
Thanks for the enhancement request @lucasmellos .  Are you asking for an "audit" of changes made by Kiali?  Or possibly an automatic update of definitions in a repository?

**Comment by leandroberetta** [comment]
Hi @lucasmellos, so basically this would be for a resource backup, or are you also using that repository for deploying other enviroments (for example using GitOps)?.

This is an interesting enhancement, recently we added the "Preview" mode where Kiali shows to the user what is going to create in an editor, and in that moment the user have also the chance to copy the resources. Of course it's a manual step by now, but maybe helps.

**Comment by lucasmellos** [comment]
Hi @leandroberetta It would be a GitOps integration 100%. For sure this preview steps would help, but considering that we're doing a high number of deployments daily it would be a nightmare handling this. 

The idea is to delegate routing capabilities to the Devs in terms of defining routing distribution and things like this.

I'll give you a complete example to help clarify things.

Let's say that I've one repo that has all the definitions i.e. (Gateway, VS, DR, etc) and the source code i.e. (Nodejs app) and as a developer a just release a new version called v-Alpha. 

I don't want v-Alpha to get http traffic from the beginning and I'd like to use Kiali to put some low percentage of this traffic to the v-Alpha. One I had this small portion of traffic to v-Alpha it would be nice to have GitOps integration from Kiali to Git. 

There're some points that it would be useful. If two develops wants to test theirs version at same time not being aware one of the other it would help a lot. 

As soon the develop 2 deploys his version from CI/CD it would override the changes made by the first developer and those change weren't tracked by any source code tool. 

In general terms it would be definitions sync to a Git repo. 

Another short term alternative could be a post to remote URL then anyone would be able to handle this on their own.

**Comment by leandroberetta** [comment]
Hi @lucasmellos thank you for the detailed explanation, I guess that posting to a remote URL would be fater to implement and it's more versatile, I guess every user will have different requirements, similar but with subtelties.

**Comment by israel-hdez** [comment]
I was thinking a little bit on this. Updating a git repository is probably somewhat out of scope of the goals of Kiali. That said, some integration with GitOps, even if indirectly, sounds reasonable.

The suggestion about calling a remote URL (like a webhook) is what probably can give more flexibility to any wanted integrations. However, I'm wondering if it would be better to use the mechanisms that are already available in the cluster, like [admission webhooks](https://kubernetes.io/docs/reference/access-authn-authz/extensible-admission-controllers/). Also, if a 3rd party integration would already require deploying some service, I wonder if using watches to monitor for changes would also be feasible?

I'm giving those alternatives because, if people can do changes via Kiali, perhaps it is also possible to apply changes via other tooling and mechanisms that can offer better monitoring of the changes may be better options.

**Comment by jshaughn** [comment]
> I'm giving those alternatives because, if people can do changes via Kiali, perhaps it is also possible to apply changes via other tooling and mechanisms that can offer better monitoring of the changes may be better options.

I agree that a direct integration with Github is likely out of scope.  But perhaps we could expand the options available with the Preview dialog.  If it were useful and not terribly difficult that would be a good integration point.  Perhaps @lucasmellos can comment, if he is still listening on the issue.

**Comment by lucasmellos** [comment]
Hey guys. I don't think it as a direct GH integration, but a Git generic or just something similar to a webhook.

**Comment by lucasmellos** [comment]
Another idea is to have it as a sidecar. Recently I implemented something similar to store dynamically istio config files. Let's say that the Git repo is a deployment label and an application sidecar would do the job of pushing it somewhere. 

I strongly agree that the sidecar is out of kiali scope, but the ability to send a webhook to some endpoint is completely within its scope.

An AdmissionWebHook may an alternative.

**Comment by israel-hdez** [comment]
I think that even generic Git support would be somewhat big, because the structure of the repo is something that would need to be standardized or configurable (and most likely, people will want it configurable). Also, we would need a writable FS somewhere, which is something that we've moved away because of some security hardening (at the moment, the FS of the Kiali container is read-only). 

So, I think that a webhook is what I see most viable. For anything more elaborated, I think it's better that somebody from community give us a hand.

**Comment by jshaughn** [comment]
I'm closing this due to lack of upvotes. Although a lot of users have a gitops workflow, this direct integration does not see to have much momentum.
